v-next/core/src/internal/tasks/resolved-task.ts

```diff
@@ -69,24 +77,190 @@ export class ResolvedTask implements Task {
     return this.actions.length === 1 && this.actions[0].action === undefined;
   }
 
+  /**
+   * This method runs the task with the given arguments.
+   * It validates the arguments, resolves the file actions, and runs the task
+   * actions by calling them in order.
+   *
+   * @param taskArguments The arguments to run the task with.
+   * @returns The result of running the task.
+   * @throws HardhatError if the task is empty, a required parameter is missing,
+   * a parameter has an invalid type, or the file actions can't be resolved.
+   */
   public async run(taskArguments: TaskArguments): Promise<any> {
-    // TODO: Run the task
-    // - Validate the argument types
-    // - Validate that there are no missing required arguments
-    // - Resolve defaults for optional arguments
-    // - Run the tasks actions with a chain of `runSuper`s
-    console.log(`Running task "${formatTaskId(this.id)}"`);
-    for (const action of this.actions) {
-      if (action.pluginId !== undefined) {
-        console.log(
-          `  Running action from plugin ${action.pluginId}: ${action.action?.toString()}`,
+    if (this.isEmpty) {
+      throw new HardhatError(HardhatError.ERRORS.TASK_DEFINITIONS.EMPTY_TASK, {
+        task: formatTaskId(this.id),
+      });
+    }
+
+    // Normalize parameters into a single iterable
+    const allParameters: TaskParameter[] = [
+      ...this.namedParameters.values(),
+      ...this.positionalParameters,
+    ];
+
+    const providedArgumentNames = new Set(Object.keys(taskArguments));
+    for (const parameter of allParameters) {
+      const value = taskArguments[parameter.name];
+
+      this.#validateRequiredParameter(parameter, value);
+      this.#validateParameterType(parameter, value);
+
+      // resolve defaults for optional parameters
+      if (value === undefined && parameter.defaultValue !== undefined) {
+        taskArguments[parameter.name] = parameter.defaultValue;
+      }
+
+      // Remove processed parameter from the set
+      providedArgumentNames.delete(parameter.name);
+    }
+
+    // At this point, the set should be empty as all the task parameters have
+    // been processed. If there are any extra parameters, an error is thrown
+    this.#validateExtraArguments(providedArgumentNames);
+
+    const next = async (
+      nextTaskArguments: TaskArguments,
+      currentIndex = this.actions.length - 1,
+    ): Promise<any> => {
+      // The first action may be empty if the task was originally an empty task
+      const currentAction = this.actions[currentIndex].action ?? (() => {});
```

Comment:
Isn't this handled by `isEmpty` at the top? It think so.

If that's the case, I think it's better to have an `assertHardhatInvariant` and let it crash if the invariant doesn't hold.

Reply:
This is a different scenario. The `isEmpty` check at the top addresses empty tasks with just one action (the one from the empty task). However, you can override the empty task and add an action to it. In this case, you need to define an empty function to avoid errors when the override action calls `runSuper`. Does that make sense? There's a test that may clarify the scenario.

Reply:
Makes sense